pkg/services/ngalert/api/api_alertmanager_guards.go

```diff
@@ -0,0 +1,157 @@
+	if !routesEqual && currentConfig.AlertmanagerConfig.Route.Provenance != ngmodels.ProvenanceNone {
+		return fmt.Errorf("policies were provisioned and cannot be changed through the UI")
+	}
+	return nil
+}
+
+func checkTemplates(currentConfig apimodels.GettableUserConfig, newConfig apimodels.PostableUserConfig) error {
+	for name, template := range currentConfig.TemplateFiles {
+		provenance := ngmodels.ProvenanceNone
+		if prov, present := currentConfig.TemplateFileProvenances[name]; present {
+			provenance = prov
+		}
+		if provenance == ngmodels.ProvenanceNone {
+			continue // we are only interested in non none
+		}
+		found := false
+		for newName, newTemplate := range newConfig.TemplateFiles {
+			if name != newName {
+				continue
+			}
+			found = true
+			if template != newTemplate {
+				return fmt.Errorf("cannot save provisioned template '%s'", name)
+			}
+			break // we found the template and we can proceed
+		}
+		if !found {
+			return fmt.Errorf("cannot delete provisioned template '%s'", name)
+		}
+	}
+	return nil
+}
+
+func checkContactPoints(current []*apimodels.GettableApiReceiver, new []*apimodels.PostableApiReceiver) error {
+	for _, existingReceiver := range current {
+		for _, existingContactPoint := range existingReceiver.GrafanaManagedReceivers {
+			if existingContactPoint.Provenance == ngmodels.ProvenanceNone {
+				continue // we are only interested in non none
+			}
+			found, edited := false, false
+		outer:
+			for _, postedReceiver := range new {
+				for _, postedContactPoint := range postedReceiver.GrafanaManagedReceivers {
+					if existingContactPoint.UID != postedContactPoint.UID {
+						continue
+					}
+					found = true
+					if existingContactPoint.DisableResolveMessage != postedContactPoint.DisableResolveMessage {
+						edited = true
+					}
+					if existingContactPoint.Name != postedContactPoint.Name {
+						edited = true
+					}
+					if existingContactPoint.Type != postedContactPoint.Type {
+						edited = true
+					}
+					for key := range existingContactPoint.SecureFields {
+						if value, present := postedContactPoint.SecureSettings[key]; present && value != "" {
+							edited = true
+							break // it's enough to know that something was edited
+						}
+					}
+					existingSettings, err := existingContactPoint.Settings.Map()
+					if err != nil {
+						return err
+					}
+					newSettings, err := postedContactPoint.Settings.Map()
+					if err != nil {
+						return err
+					}
+					for key, val := range existingSettings {
+						if newVal, present := newSettings[key]; present {
+							if val != newVal {
+								edited = true
+								break // it's enough to know that something was edited
+							}
+						} else {
+							edited = true
+							break // it's enough to know that something was edited
+						}
+					}
+					if edited {
+						return fmt.Errorf("cannot save provisioned contact point '%s'", existingContactPoint.Name)
+					}
+					break outer
```

Comment:
I don't understand why we `break outer` here. At this point only a single contact point has been found to be unedited. Don't you want to check them all?

Why don't you break as soon as edited becomes true instead of falling through and checking?

(edit: see my bigger question about structure)

Reply:
changed (using two maps now)